File: Sort/src/algorithms/quickSort.java
package algorithms;

import java.util.ArrayList;

public class quickSort {
    public static ArrayList<Integer>array;

    public static ArrayList<Integer> sort(ArrayList<Integer>array, int begin, int end){
        if (begin<end) {
            int pivot = partition(begin, end, array);

            array = sort(array, begin, pivot - 1);
            array = sort(array, pivot + 1, end);
        }

        return array;
    }

    private static int partition(int begin, int end, ArrayList<Integer>arr){
        int pivot = arr.get(end);
        int i = begin-1;

        for(int j=begin; j<end-1; j++){
            if (arr.get(j)<=pivot){
                i++;
                if(i!=j)arr = swap(arr,i,j);
            }
        }

        arr = swap(arr, i+1, end);

        array = arr;

        pivot = i+1;

        return pivot;
    }


    private static ArrayList<Integer> swap(ArrayList<Integer>array, int i, int j){
        int aux = array.get(i);
        array.set(i, array.get(j));
        array.set(j, aux);
        return array;
    }
}

File: Sort/Python/quickSort.py
# QuickSort selects a pivot and sets its greater values at its right and lower at its left
# After that, it does the same action with each side of the pivot

def quickSort(arr, begin,end):
    if begin<end:
        print()
        # Step 1
        pivot = partition(arr, begin, end)

        # Step 2
        quickSort(arr, begin, pivot-1)
        # Step 3
        quickSort(arr, pivot+1, end)


def partition(arr, begin, end):
    # Sets last element as pivot..
    i = begin-1
    pivot = arr[end]
    print(arr, pivot, begin, end)
    
    for j in range(begin,end):
        # If the element is less than the pivot, sum 1 to i, and swap it with i    
        if arr[j]<pivot:
            i+=1
            arr[i],arr[j] = arr[j], arr[i]
            print(arr)

    # Put the pivot in its right order (i+1)
    arr[i+1], arr[end] = arr[end], arr[i+1]
    print(arr)

    return i+1


arr = [5,1,3,2,6,7,5,4]
quickSort(arr, 0, len(arr)-1)
print(arr)

